drivers: dai: add DAI driver for Intel UAOL

tlissows opened the pull request:
This PR adds a DAI driver for USB Audio Offload Link (UAOL) on Intel ACE2.0 and ACE3.0 platforms.

kv2019i commented on drivers/dai/intel/uaol/uaol_regs_ace1x.h:
```diff
@@ -0,0 +1,204 @@
+#define STRMzTXDA_OFFSET(z)			(0x000 + 4 * (z))
+#define STRMzRXDA_OFFSET(z)			(0x100 + 4 * (z))
+
+union UAOLCTL {
```
This seems to be a bit excessive. There seems to be only difference in this UAOLCTL struct and UAOLxPCMSyCM_OFFSET -- all in all ~10 definitions, and rest of the 200 lines of defs are the same.

kv2019i reviewed (dismissed):
That's a big PR to go through. Some minor comments inline, but nothing to block merge. The first commit affects other DAI users, so I added Daniel to review list.

tlissows commented on drivers/dai/intel/uaol/uaol.c:
```diff
@@ -0,0 +1,1334 @@
+			break;
+		}
+		if (sys_clock_cycle_get_64() - start_time > timeout) {
+			return -ETIME;
```
Done.

tlissows commented on drivers/dai/intel/uaol/uaol.c:
```diff
@@ -0,0 +1,1334 @@
+				break;
+			}
+			if (sys_clock_cycle_get_64() - start_time > timeout) {
+				return -ETIME;
```
Done.

tlissows commented on drivers/dai/intel/uaol/uaol.c:
```diff
@@ -0,0 +1,1334 @@
+			break;
+		}
+		if (sys_clock_cycle_get_64() - start_time > timeout) {
+			return -ETIME;
```
Done.

kv2019i reviewed (commented):
Thanks @tlissows , looks good now! One question/comment I missed yesterday, can you check (see inline)?

marc-hb commented on drivers/dai/intel/uaol/CMakeLists.txt:
```diff
@@ -0,0 +1,4 @@
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_library()
```
This looks really weird. How did you test this?

tlissows commented on include/zephyr/drivers/dai.h:
```diff
@@ -113,6 +113,7 @@ enum dai_type {
 	DAI_INTEL_DMIC_NHLT,	/**< nhlt ssp */
 	DAI_INTEL_HDA_NHLT,	/**< nhlt Intel HD/A */
 	DAI_INTEL_ALH_NHLT,	/**< nhlt Intel ALH */
+	DAI_INTEL_UAOL = 16,	/**< Intel UAOL */
```
Forcing explicit value to `DAI_INTEL_UAOL` item to keep it aligned to its counterpart `SOF_DAI_INTEL_UAOL` being added to SOF. The actual definitions of `enum dai_type` (from Zephyr) and `enum sof_ipc_dai_type` (from SOF) are divergent, including a different items count.

kv2019i replied:
@tlissows Good catch. I think this is dangerous, I don't think we can assume the two enumertaions to stay aligned, and if we force one value to be guaranteed to be aligned (UAOL is 16), then that will just add to the confusion as it's very hard to mistake all types are aligned between SOF and Zephyr while in practise only UAOL is aligned.

So better to have explicit conversion from dai_config to sof_dai_config in SOF (like in dai-zephyr.c:dai_set_config).

tlissows replied:
Followed your suggestion:
- removed forcing `DAI_INTEL_UAOL` to explicit value
- rearranged SOF code in order it no longer rely on `enum dai_type` being aligned to `enum sof_ipc_dai_type` (added explicit conversion).

andyross commented on drivers/dai/intel/uaol/uaol.c:
```diff
@@ -0,0 +1,1328 @@
+ * On success, the buffer position is updated to point to the next possible TLV.
+ */
+static struct intel_uaol_ioctl_tlv *uaol_get_ioctl_tlv(uint8_t *buf, size_t buf_size,
+						       size_t *buf_pos)
```
This appears to be implementing a wire protocol.  That's a pretty weird thing to be happening at the DAI level.   Doesn't this belong in the app somewhere, which translates to the C API exposed by the driver?

It seems like there's some layers of abstraction needed here.  Where is this protocol defined?  Who else implements it (I'm guessing something in the host kernel)?

tlissows replied:
Thanks a lot for the comments! This code implements the IPC4_DMA_CONTROL protocol which is actually based on the series of TLV and needs some kind of parsing logic. Processing this at DAI level seemed reasonable to me (at least when it comes to Intel DAIs) as Intel DAIs are apparently expected to handle with IPC4 directly. I would like to stay with TLV processing at DAI level but with no ioctl extension for DAI interface.

andyross reviewed (commented):
Really not loving the ioctl mechanism here, that kind of API tends to get weird really fast in the best of cases, and in this particular case it looks like it's trying to implement some kind of host/firmware IPC, which really doesn't belong in a driver at all.

It's true that "DAI" is sort of its own device model and doesn't have a lot of clear rules, but if the DAI framework as it exists doesn't fit the needs of a driver, I think the right solution should be to write a **Zephyr** device with a Zephyr API, and have a mechanism for validating the type and getting a pointer to it out of of the DAI. Then the DAI component just becomes a wrapper.

ujfalusi commented on dts/xtensa/intel/intel_adsp_ace20_lnl.dtsi:
```diff
@@ -667,4 +687,52 @@
+			link = <0>;
+			stream = <5>;
+			status = "okay";
+		};
```
These should be a child of `uaollink0` node.
From the DAI node the link device is the parent, so you don't need to fiddle with the link properties.

tlissows replied:
Rearranged the code so that DAI nodes are children of UAOL link node(s). Thanks!

ujfalusi commented on dts/xtensa/intel/intel_adsp_ace20_lnl.dtsi:
```diff
@@ -667,4 +687,52 @@
+			stream = <1>;
+			status = "okay";
+		};
+		uaol15: uaol@f {
```
How would you name the node if the SoC contains multiple 'link' devices and they all contain a DAI which have reg = <0x0f> ?
It is better to use label which can scale later?

tlissows replied:
DAI `reg` entries correspond to `ipc4_connector_node_id::v_index`, so I think they are unique at system level and should not duplicate even for multiple UAOL IP instances. However redesigned labels of nodes to reflect their link and stream index, e.g. `uaol01:` denoting link=0 and stream=1.

MaureenHelm commented:
@tlissows do you plan to return to this PR?

tlissows commented:
> @tlissows do you plan to return to this PR?

Yes, I do. I'm going to update this PR this week.

tlissows committed:
drivers: uaol: add a driver for Intel UAOL IP

This adds a driver for USB Audio Offload Link (UAOL) IP
on Intel ACE2.0 and ACE3.0 platforms.

Signed-off-by: Tomasz Lissowski <tomasz.lissowski@intel.com>

tlissows committed:
drivers: dai: intel: uaol: add a driver for Intel UAOL DAI

This adds a DAI driver for USB Audio Offload Link (UAOL)
individual streams on Intel ACE2.0 and ACE3.0 platforms.

Signed-off-by: Tomasz Lissowski <tomasz.lissowski@intel.com>

tlissows commented on include/zephyr/drivers/dai.h:
```diff
@@ -536,6 +538,35 @@ static inline int dai_ts_get(const struct device *dev, struct dai_ts_cfg *cfg,
+ * @retval 0 If successful.
+ */
+static inline int dai_ioctl(const struct device *dev, const void *data, size_t data_size,
+							void *resp, size_t *resp_size)
```
Marking this as resolved as the PR no longer introduces the ioctl extension for DAI interface.

dnikodem reviewed (dismissed):
LGTM

andyross commented on drivers/uaol/uaol_intel_adsp.c:
```diff
@@ -0,0 +1,861 @@
+
+	tbdf.part.busn = bus;
+	tbdf.part.devn = device;
+	tbdf.part.fncn = function;
```
Style: struct initializers are your friend:

```c
union UAOLxTBDF tbdf = {
      .part.busn = bus,
      .part.devn = device,
      .part.fncn = function,
};
```

andyross replied:
Also same comment from elsewhere that single-register inlines are IMHO a waste of space.

tlissows replied:
Fixed this occurrence.

andyross commented on drivers/uaol/uaol_intel_adsp.c:
```diff
@@ -0,0 +1,861 @@
+
+	do {
+		/* write a 4x32-bit message chunk */
+		for (i = 0; i < 4 && dword_idx < dword_count; i++) {
```
```c
int n = MIN(4, msg_size / sizeof(msg[0]));
for(int i = 0; i < len; i++) {
    sys_write32(msg[i], UAOwhatever(...));
}
```

Or alternatively __ASSERT(msg_size <= 16) if you consider an overflow a program bug and not a runtime failure.

tlissows replied:
Rearranged the code to make it more clear.

andyross commented on drivers/uaol/uaol_intel_adsp.c:
```diff
@@ -0,0 +1,861 @@
+	do {
+		/* write a 4x32-bit message chunk */
+		for (i = 0; i < 4 && dword_idx < dword_count; i++) {
+			sys_write32(msg[dword_idx++], UAOLxICPy_ADDR(dp, i));
```
FWIW, this also has an uninitialized data bug if msg_size is not an integral multiple of 4.  It will write 0-3 bytes of data **after** the end of the passed msg array.

My guess is that that "should never happen", in which case __ASSERT() please.  Or change the API to take an number of words.  Or require that msg[] always be 16 bytes, etc...

tlissows replied:
Added `BUIL_ASSERT()` to ensure that memory objects being passed as `msg_buf` to `send_xhci_msg()` and `receive_xhci_msg()` functions are the whole multiple of 4 in size.

andyross reviewed (commented):
Some notes, though I didn't get all the way through.  Mostly about API and interop issues, obviously I'm not qualified to talk about the hardware.

andyross commented on drivers/uaol/uaol_intel_adsp.c:
```diff
@@ -0,0 +1,861 @@
+	sys_write32(ctl.full, UAOLCTL_ADDR(dp));
+
+	start_time = sys_clock_cycle_get_64();
+	while (1) {
```
(Should be "`end - k_cycle_get_32() > 0`" of course, time math is hard: no one should trust me near that stuff)

tlissows replied:
Done. Used `WAIT_FOR()` macro.

aescolar commented:
Bulk changing the milestone from everything which did not make it to the v3.7.0 freeze deadline
and which is not labeled as a bug or documentation (or does not appear clearly as only containing
a bug fix or a documentation change).
If you disagree about this change please add the milestone again.

Please note that until rc2 only PRs containing bug fixes, docs and tests can be merged.
After that and until release only bug fixes and docs changes.

Exceptions require approval by the release team and a vote by the TSC.
https://docs.zephyrproject.org/latest/project/release_process.html#stabilization-phase

tlissows commented on drivers/uaol/uaol_intel_adsp.c:
```diff
@@ -0,0 +1,861 @@
+		ir.part.irvi = 1;
+		sys_write32(ir.full, UAOLxIR_ADDR(dp));
+
+		/* loop more if HW anticipates more payload and there's free space in buffer */
```
Done.

tlissows commented on drivers/uaol/uaol_intel_adsp.c:
```diff
@@ -0,0 +1,861 @@
+	ctl.part.spa = (power) ? (ctl.part.spa | link_mask) : (ctl.part.spa & ~link_mask);
+	sys_write32(ctl.full, UAOLCTL_ADDR(dp));
+
+	start_time = sys_clock_cycle_get_64();
```
Used `k_cycle_get` instead of `sys_clock_cycle_get`. Also switched from 64-bit to 32-bit math. Changes made in all places where applicable. Thanks !

tlissows commented on drivers/uaol/uaol_intel_adsp.c:
```diff
@@ -0,0 +1,861 @@
+{
+	struct uaol_intel_adsp_data *dp = dev->data;
+
+	sys_write16(channel_map, UAOLxPCMSyCM_ADDR(dp, stream));
```
Removed `program_fifo_start_offset` and `program_channel_map` functions. Put `sys_write16` inline where it's being called.

tlissows commented on drivers/uaol/uaol_intel_adsp.c:
```diff
@@ -0,0 +1,861 @@
+
+static int uaol_intel_adsp_lock(const struct device *dev)
+{
+	return k_mutex_lock(&mutex, K_FOREVER);
```
It wasn't my intention to introduce `uaol_lock()/uaol_unlock()` API functions. Rather wanted to add `lock/unlock` pointers to get compact way to use for the actual API functions. Such an approach can be found in usb\udc.h or usb\uhc.h. Dispensing with this approach as potentially confusing.

tmleman commented on drivers/dai/intel/uaol/uaol.c:
```diff
@@ -0,0 +1,416 @@
+
+	return 0;
+}
+
+static int dai_uaol_config_get(const struct device *dev, struct dai_config *cfg, enum dai_dir dir)
+{
```
`dir` is an unused argument. To prevent unnecessary compiler warnings and clearly indicate that this parameter is intentionally unused, please apply the `ARG_UNUSED` macro.

tlissows replied:
Fixed. Thanks !

tmleman commented on drivers/dai/intel/uaol/uaol.c:
```diff
@@ -0,0 +1,416 @@
+
+	return 0;
+}
+
+static const struct dai_properties *dai_uaol_get_properties(const struct device *dev,
+							    enum dai_dir dir, int stream_id)
+{
```
ditto for `stream_id`  (and `dir` on non-MTL platforms).

tlissows replied:
Fixed.

tmleman commented on drivers/dai/intel/uaol/uaol.c:
```diff
@@ -0,0 +1,416 @@
+#endif
+	return prop;
+}
+
+static int dai_uaol_trigger(const struct device *dev, enum dai_dir dir, enum dai_trigger_cmd cmd)
+{
```
ditto

tlissows replied:
Fixed.

tmleman commented on drivers/dai/intel/uaol/uaol-params-intel-ipc4.h:
```diff
@@ -0,0 +1,100 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __INTEL_DAI_DRIVER_UAOL_PARAMS_IPC4_H__
+#define __INTEL_DAI_DRIVER_UAOL_PARAMS_IPC4_H__
```
Why not `INTEL_DAI_DRIVER_UAOL_PARAMS_IPC4_H_`?

tlissows replied:
Fixed.

tmleman reviewed (changes requested):
Please prepare a pull request in SOF that points to this PR. Rebuilding this with SOF will avoid the need to return here with minor fixes after the merge. Currently, I am not sure, but previously in some SOF workflows, warnings were treated as errors.

tmleman reviewed (dismissed):
The solution of moving the driver for UAOL to a lower level than DAI is okay. However, you have device power management implemented twice because of this. Is there a reason for this? In other DAIs, we have device power management because there is no driver below for the given interface. I think power management in `drivers/uaol/uaol_intel_adsp.c` would be sufficient. Additionally, the appropriate power domains should be added to the DTS (UAOL is in the `hst_domain` domain).

tlissows commented on drivers/dai/intel/uaol/uaol.c:
```diff
@@ -0,0 +1,416 @@
+	ret = dai_uaol_parse_aux_data(dev, bespoke_cfg);
+	if (ret) {
+		return ret;
+	}
```
Currently, the whole support for MTL is removed, so the remark does not apply anymore.

tlissows commented:
> The solution of moving the driver for UAOL to a lower level than DAI is okay. However, you have device power management implemented twice because of this. Is there a reason for this? In other DAIs, we have device power management because there is no driver below for the given interface. I think power management in `drivers/uaol/uaol_intel_adsp.c` would be sufficient. Additionally, the appropriate power domains should be added to the DTS (UAOL is in the `hst_domain` domain).

UAOL power domains added in DTS. As for power management, exposing it by UAOL DAI driver (in addition to UAOL IP driver) seems reasonable due to its ease of use on application side. As an example please see basefw_vendor_dma_control() function (in SOF), where you can find power mnanagement API called directly for DAI devices. Calls for UAOL IP devices would also be possible, but much more complicated. Some extra steps would be needed (like UAOL DAI -> IP mapping) thus creating an exceptional path for UAOL.

tlissows commented:
Removed support for ACE1.x platform from this PR. Also aligned to the changes introduced by commit [include: dai: Introduce runtime DAI configuration update API](https://github.com/zephyrproject-rtos/zephyr/pull/73158/commits/f1eb0402536fbf41c09df6e49b7d40b27a640699).

tmleman reviewed (dismissed):
My earlier comments have been addressed.

tlissows commented:
Rebased on top of main.
Used WAIT_FOR() macro where applicable.
